applications/CompressiblePotentialFlowApplication/custom_elements/embedded_incompressible_potential_flow_element.cpp

```diff
@@ -0,0 +1,184 @@
+        NewId, pGeom, pProperties);
+    KRATOS_CATCH("");
+}
+
+template <int Dim, int NumNodes>
+Element::Pointer EmbeddedIncompressiblePotentialFlowElement<Dim, NumNodes>::Clone(
+    IndexType NewId, NodesArrayType const& ThisNodes) const
+{
+    KRATOS_TRY
+    return Kratos::make_shared<EmbeddedIncompressiblePotentialFlowElement>(
+        NewId, this->GetGeometry().Create(ThisNodes), this->pGetProperties());
+    KRATOS_CATCH("");
+}
+
+template <int Dim, int NumNodes>
+void EmbeddedIncompressiblePotentialFlowElement<Dim, NumNodes>::CalculateLocalSystem(
+    MatrixType& rLeftHandSideMatrix, VectorType& rRightHandSideVector, ProcessInfo& rCurrentProcessInfo)
+{
+    const EmbeddedIncompressiblePotentialFlowElement& r_this = *this;
+    const int wake = r_this.GetValue(WAKE);
+    const int kutta = r_this.GetValue(KUTTA);
+
+
+    if (this->Is(BOUNDARY) && wake == 0 && kutta == 0)
+        CalculateEmbeddedLocalSystem(rLeftHandSideMatrix,rRightHandSideVector,rCurrentProcessInfo);
+    else
+        BaseType::CalculateLocalSystem(rLeftHandSideMatrix, rRightHandSideVector, rCurrentProcessInfo);
+
+}
+
+template <int Dim, int NumNodes>
+void EmbeddedIncompressiblePotentialFlowElement<Dim, NumNodes>::CalculateEmbeddedLocalSystem(
+    MatrixType& rLeftHandSideMatrix, VectorType& rRightHandSideVector, ProcessInfo& rCurrentProcessInfo)
+{
+    if (rLeftHandSideMatrix.size1() != NumNodes || rLeftHandSideMatrix.size2() != NumNodes)
+        rLeftHandSideMatrix.resize(NumNodes, NumNodes, false);
+    if (rRightHandSideVector.size() != NumNodes)
+        rRightHandSideVector.resize(NumNodes, false);
+    rLeftHandSideMatrix.clear();
+
+    array_1d<double, NumNodes> potential;
+    Vector distances(NumNodes);
+    for(unsigned int i_node = 0; i_node<NumNodes; i_node++)
+        distances(i_node) = this->GetGeometry()[i_node].GetSolutionStepValue(LEVEL_SET);
+
+    BaseType::GetPotentialOnNormalElement(potential);
+
+    ModifiedShapeFunctions::Pointer pModifiedShFunc = this->pGetModifiedShapeFunctions(distances);
+    Matrix positive_side_sh_func;
+    ModifiedShapeFunctions::ShapeFunctionsGradientsType positive_side_sh_func_gradients;
+    Vector positive_side_weights;
+    pModifiedShFunc -> ComputePositiveSideShapeFunctionsAndGradientsValues(
+        positive_side_sh_func,
+        positive_side_sh_func_gradients,
+        positive_side_weights,
+        GeometryData::GI_GAUSS_2);
+
+        BoundedMatrix<double,NumNodes,NumNodes> aux_matrix;
+        BoundedMatrix<double,NumNodes,Dim> DN_DX;
+        for (unsigned int i_gauss=0;i_gauss<positive_side_sh_func_gradients.size();i_gauss++){
+            DN_DX=positive_side_sh_func_gradients(i_gauss);
+            aux_matrix=prod(DN_DX,trans(DN_DX))*positive_side_weights(i_gauss);
+
+            noalias(rLeftHandSideMatrix) += aux_matrix;
+        }
+
+    noalias(rRightHandSideVector) = -prod(rLeftHandSideMatrix, potential);
+}
+
+template <int Dim, int NumNodes>
```

Comment:
```suggestion
template <>
ModifiedShapeFunctions::Pointer EmbeddedIncompressiblePotentialFlowElement<2,3>::pGetModifiedShapeFunctions(Vector& rDistances) {                   
    return Kratos::make_shared<Triangle2D3ModifiedShapeFunctions>(this->pGetGeometry(), rDistances);
}
```
You can specialize the function for the 2D and 3D cases so you avoid all the if statements (this is the idea). As a comment aside, all pointers should be created with the Kratos `Kratos::make_shared<Type>(constructor args)` function.

Reply:
Actually I think that this one can be a unique pointer.